library/ziti.c

```diff
@@ -439,3 +724,160 @@ static void version_cb(ctrl_version *v, ziti_error *err, void *ctx) {
         free_ctrl_version(v);
     }
 }
+
+#define OID_PKCS7 MBEDTLS_OID_PKCS "\x07"
+#define OID_PKCS7_DATA OID_PKCS7 "\x02"
+#define OID_PKCS7_SIGNED_DATA OID_PKCS7 "\x01"
+
+static void well_known_certs_cb(char *base64_encoded_pkcs7, ziti_error *err, void *req) {
+    struct nf_enroll_req *enroll_req = req;
+    size_t der_len;
+    unsigned char *p;
+    unsigned char *end;
+    mbedtls_x509_crt *cp;
+    mbedtls_x509_crt certs={0};
+    unsigned char *cert_buf;
+
+    int rc = mbedtls_base64_decode( NULL, 0, &der_len, base64_encoded_pkcs7, strlen(base64_encoded_pkcs7)); // determine necessary buffer size
+    char *base64_decoded_pkcs7 = calloc(1, der_len + 1);
+    rc = mbedtls_base64_decode( base64_decoded_pkcs7, der_len, &der_len, base64_encoded_pkcs7, strlen(base64_encoded_pkcs7));
+
+    unsigned char *der = (unsigned char*)base64_decoded_pkcs7;
+
+    p = der;
+    end = der + der_len;
+    size_t len;
+    mbedtls_asn1_get_tag(&p, end, &len, MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE);
+
+    mbedtls_asn1_get_tag(&p, end, &len, MBEDTLS_ASN1_OID);
+
+    mbedtls_asn1_buf oid;
+    oid.p = p;
+    oid.len = len;
+    if (!MBEDTLS_OID_CMP(OID_PKCS7_SIGNED_DATA, &oid)) {
+        ZITI_LOG(ERROR, "invalid pkcs7 signed data");
+        return;
+    }
+    p += len;
+
+    mbedtls_asn1_get_tag(&p, end, &len, MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_CONTEXT_SPECIFIC);
+
+    mbedtls_asn1_get_tag(&p, end, &len, MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE);
+
+    int ver;
+    mbedtls_asn1_get_int(&p, end, &ver);
+
+    mbedtls_asn1_get_tag(&p, end, &len, MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SET);
+    mbedtls_asn1_get_tag(&p, end, &len, MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE);
+
+    mbedtls_asn1_get_tag(&p, end, &len, MBEDTLS_ASN1_OID);
+    oid.p = p;
+    oid.len = len;
+    if (!MBEDTLS_OID_CMP(OID_PKCS7_DATA, &oid)) {
+        ZITI_LOG(ERROR, "invalid pkcs7 data");
+        return;
+    }
+    p += len;
+
+    mbedtls_asn1_get_tag(&p, end, &len, MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_CONTEXT_SPECIFIC);
+
+    cert_buf = p;
+
+    do {
+        size_t cert_len;
+        unsigned char *cbp = cert_buf;
+        rc = mbedtls_asn1_get_tag(&cbp, end, &cert_len, MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE);
+        if (rc != 0) break;
+        cert_len += (cbp - cert_buf);
+        rc = mbedtls_x509_crt_parse(&certs, cert_buf, cert_len);
+        cert_buf += cert_len;
+    } while (rc == 0);
+
+    cp = &certs;
+
+    do {
+
+        size_t olen;
+        mbedtls_base64_encode( NULL, 0, &olen, cp->raw.p, cp->raw.len );  // determine size of buffer we need to allocate
+        enroll_req->ecfg->CA = calloc(1, olen + 1);
```

Comment:
you may get more than one (in the loop), this will override and keep the last one. Rather get the whole chain and write it out in the end

Reply:
good catch.  I now create/manage a list of certs and combine/emit all of them in the chain.